File: MediaMonitor.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Windows.Media.Control;
using Windows.Media;

namespace MediaTracker
{
    public class MediaMonitor : IDisposable
    {
        // whitelist ключевых слов для музыкальных приложений (проверяем SourceAppUserModelId)
        private static readonly string[] MusicAppKeywords =
        [
            "spotify",
            "yandex",
            "yandexmusic",
            "yamusic",
            "yandex.music"
        ];

        // blacklist ключевых слов для явно не-музыкальных источников (встречаются в title или id)
        private static readonly string[] NonMusicKeywords =
        [
            "twitch",
            "youtube",
            "netflix",
            "prime video",
            "primevideo",
            "mozilla",
            "firefox",
            "chrome",
            "edge",
            "vimeo",
            "soundcloud", // soundcloud — может быть муз., но при желании убрать
            "facebook",
            "instagram"
        ];

        private readonly object _lock = new();
        private GlobalSystemMediaTransportControlsSessionManager? _manager;
        private GlobalSystemMediaTransportControlsSession? _session;
        private CancellationTokenSource? _backgroundCts;
        private System.Threading.Timer? _debounceTimer;
        private List<GlobalSystemMediaTransportControlsSession> _subscribedSessions = new();

        private MediaState _current = new();
        public MediaState Current
        {
            get { lock (_lock) return _current; }
            private set { lock (_lock) { _current = value; } }
        }

        public event Action<MediaState>? StateChanged;
        public bool IsStarted { get; private set; } = false;

        public async Task StartAsync(CancellationToken token = default)
        {
            if (IsStarted) return;
            IsStarted = true;

            _backgroundCts = CancellationTokenSource.CreateLinkedTokenSource(token);

            _manager = await GlobalSystemMediaTransportControlsSessionManager.RequestAsync();

            _manager.SessionsChanged += Manager_SessionsChanged;
            _manager.CurrentSessionChanged += Manager_CurrentSessionChanged;

            UpdateCurrentSession();
            ScheduleRefresh(0);
        }

        public async Task StopAsync()
        {
            if (!IsStarted) return;
            IsStarted = false;

            try
            {
                _debounceTimer?.Change(Timeout.Infinite, Timeout.Infinite);
                _backgroundCts?.Cancel();

                lock (_lock)
                {
                    if (_session != null)
                    {
                        try
                        {
                            _session.MediaPropertiesChanged -= Session_MediaPropertiesChanged;
                            _session.PlaybackInfoChanged -= Session_PlaybackInfoChanged;
                        }
                        catch { }
                    }

                    if (_manager != null)
                    {
                        try
                        {
                            _manager.SessionsChanged -= Manager_SessionsChanged;
                            _manager.CurrentSessionChanged -= Manager_CurrentSessionChanged;
                        }
                        catch { }
                    }

                    // Unsubscribe from all sessions
                    foreach (var s in _subscribedSessions)
                    {
                        try
                        {
                            s.MediaPropertiesChanged -= Session_MediaPropertiesChanged;
                            s.PlaybackInfoChanged -= Session_PlaybackInfoChanged;
                        }
                        catch { }
                    }
                    _subscribedSessions.Clear();
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"MediaMonitor.StopAsync exception: {ex}");
            }

            await Task.CompletedTask;
        }

        private void Manager_CurrentSessionChanged(GlobalSystemMediaTransportControlsSessionManager? sender, CurrentSessionChangedEventArgs? args)
            => ScheduleSessionUpdate();

        private void Manager_SessionsChanged(GlobalSystemMediaTransportControlsSessionManager? sender, object? args)
            => ScheduleSessionUpdate();

        private void ScheduleSessionUpdate()
        {
            UpdateCurrentSession();
            ScheduleRefresh();
        }

        private void UpdateCurrentSession()
        {
            try
            {
                if (_manager == null) return;
                var sessions = _manager.GetSessions().ToList();

                // Unsubscribe from old sessions
                foreach (var s in _subscribedSessions)
                {
                    try
                    {
                        s.MediaPropertiesChanged -= Session_MediaPropertiesChanged;
                        s.PlaybackInfoChanged -= Session_PlaybackInfoChanged;
                    }
                    catch { }
                }
                _subscribedSessions.Clear();

                // Subscribe to new sessions
                _subscribedSessions.AddRange(sessions);
                foreach (var s in sessions)
                {
                    s.MediaPropertiesChanged += Session_MediaPropertiesChanged;
                    s.PlaybackInfoChanged += Session_PlaybackInfoChanged;
                }

                var pick = _manager.GetCurrentSession();

                if (pick != null)
                {
                    string id = pick.SourceAppUserModelId ?? "";
                    if (!id.Contains("spotify", StringComparison.OrdinalIgnoreCase) &&
                        !id.Contains("yandex", StringComparison.OrdinalIgnoreCase))
                    {
                        pick = sessions.FirstOrDefault(s => s.SourceAppUserModelId?.Contains("spotify", StringComparison.OrdinalIgnoreCase) == true)
                               ?? sessions.FirstOrDefault(s => s.SourceAppUserModelId?.Contains("yandex", StringComparison.OrdinalIgnoreCase) == true)
                               ?? pick;
                    }
                }
                else
                {
                    pick = sessions.FirstOrDefault(s => s.SourceAppUserModelId?.Contains("spotify", StringComparison.OrdinalIgnoreCase) == true)
                           ?? sessions.FirstOrDefault(s => s.SourceAppUserModelId?.Contains("yandex", StringComparison.OrdinalIgnoreCase) == true)
                           ?? sessions.FirstOrDefault();
                }

                lock (_lock)
                {
                    if (_session != null && _session != pick)
                    {
                        try
                        {
                            _session.MediaPropertiesChanged -= Session_MediaPropertiesChanged;
                            _session.PlaybackInfoChanged -= Session_PlaybackInfoChanged;
                        }
                        catch { }
                    }

                    _session = pick;

                    if (_session != null)
                    {
                        _session.MediaPropertiesChanged += Session_MediaPropertiesChanged;
                        _session.PlaybackInfoChanged += Session_PlaybackInfoChanged;
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"UpdateCurrentSession error: {ex}");
            }
        }

        private void Session_MediaPropertiesChanged(GlobalSystemMediaTransportControlsSession? sender, MediaPropertiesChangedEventArgs? args)
            => ScheduleRefresh();

        private void Session_PlaybackInfoChanged(GlobalSystemMediaTransportControlsSession? sender, PlaybackInfoChangedEventArgs? args)
            => ScheduleRefresh();

        private void ScheduleRefresh(int debounceMs = 150)
        {
            lock (_lock)
            {
                _debounceTimer?.Change(Timeout.Infinite, Timeout.Infinite);
                if (_debounceTimer == null)
                {
                    _debounceTimer = new System.Threading.Timer(_ => _ = RefreshNowPlayingAsync(), null, debounceMs, Timeout.Infinite);
                }
                else
                {
                    _debounceTimer.Change(debounceMs, Timeout.Infinite);
                }
            }
        }

        private async Task RefreshNowPlayingAsync()
        {
            try
            {
                if (_manager == null)
                {
                    PublishStateIfChanged(new MediaState());
                    return;
                }

                var sessions = _manager.GetSessions().ToList();
                if (!sessions.Any()) { PublishStateIfChanged(new MediaState()); return; }

                var tasks = sessions.Select(async s =>
                {
                    try
                    {
                        var propsTask = s.TryGetMediaPropertiesAsync().AsTask();
                        var completed = await Task.WhenAny(propsTask, Task.Delay(800));
                        if (completed != propsTask) return (s: s, artist: "", title: "", isPlaying: false, id: s.SourceAppUserModelId ?? "");
                        var props = await propsTask;
                        var info = s.GetPlaybackInfo();
                        string artist = props?.Artist?.Trim() ?? "";
                        string title = props?.Title?.Trim() ?? "";
                        bool isPlaying = info?.PlaybackStatus == GlobalSystemMediaTransportControlsSessionPlaybackStatus.Playing;
                        string id = s.SourceAppUserModelId ?? "";
                        return (s, artist, title, isPlaying, id);
                    }
                    catch { return (s: s, artist: "", title: "", isPlaying: false, id: s.SourceAppUserModelId ?? ""); }
                }).ToArray();

                var infos = (await Task.WhenAll(tasks)).ToList();

                // 1) Сначала — играющие сессии из whitelist (Spotify, Yandex)
                var playingWhitelisted = infos
                    .Where(x => x.isPlaying && ContainsAnyKeyword(x.id, MusicAppKeywords))
                    .OrderBy(x => !ContainsAnyKeyword(x.id, MusicAppKeywords)) // сохраняем порядок (необязательно)
                    .FirstOrDefault();

                if (playingWhitelisted.s != null)
                {
                    PublishStateIfChanged(new MediaState
                    {
                        Artist = playingWhitelisted.artist,
                        Title = playingWhitelisted.title,
                        IsPlaying = true,
                        SourceApp = playingWhitelisted.id,
                        Timestamp = DateTimeOffset.UtcNow
                    });
                    return;
                }

                // 2) Если нет — ищем играющие сессии, которые выглядят как музыка:
                //    - есть artist OR title выглядит не как страница (не содержит blacklist)
                var playingLikelyMusic = infos
                    .Where(x => x.isPlaying && ( !string.IsNullOrWhiteSpace(x.artist) || !ContainsAnyKeyword(x.title + x.id, NonMusicKeywords) ))
                    .OrderBy(x => string.IsNullOrWhiteSpace(x.artist) ? 1 : 0) // предпочитаем с artist
                    .FirstOrDefault();

                if (playingLikelyMusic.s != null)
                {
                    PublishStateIfChanged(new MediaState
                    {
                        Artist = playingLikelyMusic.artist,
                        Title = playingLikelyMusic.title,
                        IsPlaying = true,
                        SourceApp = playingLikelyMusic.id,
                        Timestamp = DateTimeOffset.UtcNow
                    });
                    return;
                }

                // 3) Если всё ещё ничего — (опционально) считать другие playing как fallback,
                //    но фильтруем по blacklist (не показываем Twitch/YouTube если можно)
                var playingOther = infos
                    .Where(x => x.isPlaying && !ContainsAnyKeyword(x.title + x.id, NonMusicKeywords))
                    .FirstOrDefault();

                if (playingOther.s != null)
                {
                    PublishStateIfChanged(new MediaState
                    {
                        Artist = playingOther.artist,
                        Title = playingOther.title,
                        IsPlaying = true,
                        SourceApp = playingOther.id,
                        Timestamp = DateTimeOffset.UtcNow
                    });
                    return;
                }

                // 4) Ничего подходящего не играет — показываем лучший last-known (как раньше)
                var currentSession = _manager.GetCurrentSession();
                var displayPick = infos.FirstOrDefault(x => x.s == currentSession);
                if (displayPick.s == null)
                {
                    displayPick = infos
                        .OrderBy(x =>
                        {
                            if (ContainsAnyKeyword(x.id, MusicAppKeywords)) return 0;
                            if (!string.IsNullOrWhiteSpace(x.artist)) return 1;
                            return 2;
                        })
                        .FirstOrDefault();
                }

                if (displayPick.s != null)
                {
                    PublishStateIfChanged(new MediaState
                    {
                        Artist = displayPick.artist,
                        Title = displayPick.title,
                        IsPlaying = false,
                        SourceApp = displayPick.id,
                        Timestamp = DateTimeOffset.UtcNow
                    });
                }
                else
                {
                    PublishStateIfChanged(new MediaState());
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"RefreshNowPlayingAsync (aggregate) error: {ex}");
                PublishStateIfChanged(new MediaState());
            }
        }

        private void PublishStateIfChanged(MediaState newState)
        {
            bool changed = false;
            lock (_lock)
            {
                var prev = _current;
                if (!string.Equals(prev.Artist, newState.Artist, StringComparison.Ordinal)
                    || !string.Equals(prev.Title, newState.Title, StringComparison.Ordinal)
                    || prev.IsPlaying != newState.IsPlaying
                    || !string.Equals(prev.SourceApp, newState.SourceApp, StringComparison.Ordinal))
                {
                    _current = newState;
                    changed = true;
                }
            }
            if (changed)
            {
                try { StateChanged?.Invoke(newState); } catch { }
            }
        }

        public async Task<bool> TogglePlayPauseAsync()
        {
            var state = Current;
            var sessions = _manager?.GetSessions() ?? Enumerable.Empty<GlobalSystemMediaTransportControlsSession>();
            var session = sessions.FirstOrDefault(s => string.Equals(s.SourceAppUserModelId, state.SourceApp, StringComparison.OrdinalIgnoreCase))
                          ?? _session;
            if (session == null) return false;

            try
            {
                var info = session.GetPlaybackInfo();
                if (info?.PlaybackStatus == GlobalSystemMediaTransportControlsSessionPlaybackStatus.Playing)
                    await session.TryPauseAsync();
                else
                    await session.TryPlayAsync();

                ScheduleRefresh();
                return true;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"TogglePlayPauseAsync failed: {ex}");
                return false;
            }
        }

        public void Dispose()
        {
            try { _debounceTimer?.Change(Timeout.Infinite, Timeout.Infinite); } catch { }
            try { _backgroundCts?.Cancel(); } catch { }
            try { _ = StopAsync(); } catch { }
        }

        private static bool ContainsAnyKeyword(string text, string[] keywords)
        {
            if (string.IsNullOrWhiteSpace(text)) return false;
            foreach (var k in keywords)
                if (text.IndexOf(k, StringComparison.OrdinalIgnoreCase) >= 0)
                    return true;
            return false;
        }
    }
}

File: src/media-monitor.cpp
#include "media-monitor.hpp"

#include <algorithm>
#include <array>
#include <chrono>
#include <cwctype>
#include <utility>
#include <thread>
#include <vector>
#include <winrt/Windows.Foundation.Collections.h>
#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.Media.Control.h>

namespace kamyshtracker {
namespace {

using winrt::Windows::Media::Control::GlobalSystemMediaTransportControlsSession;
using winrt::Windows::Media::Control::GlobalSystemMediaTransportControlsSessionManager;
using winrt::Windows::Media::Control::GlobalSystemMediaTransportControlsSessionPlaybackStatus;

constexpr auto fallbackRefreshInterval = std::chrono::seconds(45);
constexpr auto eventCoalesceDelay = std::chrono::milliseconds(350);

constexpr std::array<std::wstring_view, 5> musicKeywords = {
    L"spotify", L"yandex", L"yandexmusic", L"yamusic", L"yandex.music"
};

constexpr std::array<std::wstring_view, 14> nonMusicKeywords = {
    L"twitch", L"youtube", L"netflix", L"prime video", L"primevideo", L"mozilla",
    L"firefox", L"chrome", L"edge", L"vimeo", L"soundcloud", L"facebook",
    L"instagram", L"browser"
};

std::wstring lower(std::wstring value)
{
    std::transform(value.begin(), value.end(), value.begin(), [](wchar_t c) {
        return static_cast<wchar_t>(std::towlower(c));
    });
    return value;
}

bool containsAny(const std::wstring &text, const auto &keywords)
{
    const auto lowered = lower(text);
    for (const auto keyword : keywords) {
        if (lowered.find(keyword) != std::wstring::npos)
            return true;
    }
    return false;
}

struct SessionInfo {
    GlobalSystemMediaTransportControlsSession session{nullptr};
    std::wstring artist;
    std::wstring title;
    bool isPlaying = false;
    std::wstring id;
};

MediaState toState(const SessionInfo &info, bool isPlaying)
{
    return MediaState{
        .artist = info.artist,
        .title = info.title,
        .isPlaying = isPlaying,
        .sourceApp = info.id,
        .timestamp = std::chrono::system_clock::now(),
    };
}

void clearSubscriptions(std::vector<MediaSessionSubscription> &subscriptions)
{
    for (auto &subscription : subscriptions) {
        try {
            if (subscription.session) {
                subscription.session.MediaPropertiesChanged(subscription.mediaPropertiesChanged);
                subscription.session.PlaybackInfoChanged(subscription.playbackInfoChanged);
            }
        } catch (...) {
        }
    }
    subscriptions.clear();
}

}

MediaMonitor::MediaMonitor() = default;

MediaMonitor::~MediaMonitor()
{
    stop();
}

bool MediaMonitor::start()
{
    bool expected = false;
    if (!running_.compare_exchange_strong(expected, true))
        return true;

    {
        std::lock_guard lock(wakeMutex_);
        refreshRequested_ = true;
    }
    worker_ = std::thread([this] { workerLoop(); });
    return true;
}

void MediaMonitor::stop()
{
    if (!running_.exchange(false))
        return;

    wake_.notify_all();
    if (worker_.joinable())
        worker_.join();
}

MediaState MediaMonitor::current() const
{
    std::lock_guard lock(mutex_);
    return current_;
}

void MediaMonitor::setCallback(Callback callback)
{
    std::lock_guard lock(mutex_);
    callback_ = std::move(callback);
}

void MediaMonitor::workerLoop()
{
    winrt::init_apartment(winrt::apartment_type::multi_threaded);

    GlobalSystemMediaTransportControlsSessionManager manager{nullptr};
    winrt::event_token sessionsChanged{};
    winrt::event_token currentSessionChanged{};
    std::vector<MediaSessionSubscription> subscriptions;

    auto requestRefresh = [this] {
        {
            std::lock_guard lock(wakeMutex_);
            refreshRequested_ = true;
        }
        wake_.notify_one();
    };

    try {
        manager = GlobalSystemMediaTransportControlsSessionManager::RequestAsync().get();
        sessionsChanged = manager.SessionsChanged([requestRefresh](auto &&, auto &&) {
            requestRefresh();
        });
        currentSessionChanged = manager.CurrentSessionChanged([requestRefresh](auto &&, auto &&) {
            requestRefresh();
        });
    } catch (...) {
        publishIfChanged({});
    }

    auto nextFallback = std::chrono::steady_clock::now();
    while (running_) {
        bool shouldRefresh = false;
        bool eventDriven = false;
        {
            std::unique_lock lock(wakeMutex_);
            if (!refreshRequested_) {
                wake_.wait_until(lock, nextFallback, [this] {
                    return !running_ || refreshRequested_;
                });
            }

            if (!running_)
                break;

            const auto now = std::chrono::steady_clock::now();
            eventDriven = refreshRequested_;
            shouldRefresh = refreshRequested_ || now >= nextFallback;
            refreshRequested_ = false;
        }

        if (!shouldRefresh)
            continue;

        if (eventDriven) {
            std::unique_lock lock(wakeMutex_);
            wake_.wait_for(lock, eventCoalesceDelay, [this] {
                return !running_;
            });
            if (!running_)
                break;
            refreshRequested_ = false;
        }

        if (manager)
            refresh(manager, subscriptions, requestRefresh);
        else
            refresh();
        nextFallback = std::chrono::steady_clock::now() + fallbackRefreshInterval;
    }

    clearSubscriptions(subscriptions);
    try {
        if (manager) {
            manager.SessionsChanged(sessionsChanged);
            manager.CurrentSessionChanged(currentSessionChanged);
        }
    } catch (...) {
    }

    winrt::uninit_apartment();
}

void MediaMonitor::refresh()
{
    try {
        auto manager = GlobalSystemMediaTransportControlsSessionManager::RequestAsync().get();
        std::vector<MediaSessionSubscription> subscriptions;
        refresh(manager, subscriptions, {});
    } catch (...) {
        publishIfChanged({});
    }
}

void MediaMonitor::refresh(
    const GlobalSystemMediaTransportControlsSessionManager &manager,
    std::vector<MediaSessionSubscription> &subscriptions,
    const std::function<void()> &requestRefresh)
{
    try {
        auto sessions = manager.GetSessions();
        clearSubscriptions(subscriptions);

        std::vector<SessionInfo> infos;
        infos.reserve(sessions.Size());
        subscriptions.reserve(sessions.Size());

        for (const auto &session : sessions) {
            if (requestRefresh) {
                MediaSessionSubscription subscription;
                subscription.session = session;
                subscription.mediaPropertiesChanged = session.MediaPropertiesChanged([requestRefresh](auto &&, auto &&) {
                    requestRefresh();
                });
                subscription.playbackInfoChanged = session.PlaybackInfoChanged([requestRefresh](auto &&, auto &&) {
                    requestRefresh();
                });
                subscriptions.push_back(std::move(subscription));
            }

            SessionInfo info;
            info.session = session;
            info.id = std::wstring(session.SourceAppUserModelId().c_str());

            try {
                auto props = session.TryGetMediaPropertiesAsync().get();
                info.artist = std::wstring(props.Artist().c_str());
                info.title = std::wstring(props.Title().c_str());
            } catch (...) {
            }

            try {
                const auto playback = session.GetPlaybackInfo();
                info.isPlaying =
                    playback.PlaybackStatus() == GlobalSystemMediaTransportControlsSessionPlaybackStatus::Playing;
            } catch (...) {
            }

            infos.push_back(std::move(info));
        }

        if (infos.empty()) {
            publishIfChanged({});
            return;
        }

        auto playingWhitelisted = std::find_if(infos.begin(), infos.end(), [](const SessionInfo &info) {
            return info.isPlaying && containsAny(info.id, musicKeywords);
        });
        if (playingWhitelisted != infos.end()) {
            publishIfChanged(toState(*playingWhitelisted, true));
            return;
        }

        auto playingLikelyMusic = std::find_if(infos.begin(), infos.end(), [](const SessionInfo &info) {
            return info.isPlaying &&
                (!info.artist.empty() || !containsAny(info.title + info.id, nonMusicKeywords));
        });
        if (playingLikelyMusic != infos.end()) {
            publishIfChanged(toState(*playingLikelyMusic, true));
            return;
        }

        auto playingOther = std::find_if(infos.begin(), infos.end(), [](const SessionInfo &info) {
            return info.isPlaying && !containsAny(info.title + info.id, nonMusicKeywords);
        });
        if (playingOther != infos.end()) {
            publishIfChanged(toState(*playingOther, true));
            return;
        }

        auto currentSession = manager.GetCurrentSession();
        auto displayPick = std::find_if(infos.begin(), infos.end(), [&](const SessionInfo &info) {
            return info.session == currentSession;
        });

        if (displayPick == infos.end()) {
            std::stable_sort(infos.begin(), infos.end(), [](const SessionInfo &left, const SessionInfo &right) {
                const auto rank = [](const SessionInfo &info) {
                    if (containsAny(info.id, musicKeywords))
                        return 0;
                    if (!info.artist.empty())
                        return 1;
                    return 2;
                };
                return rank(left) < rank(right);
            });
            displayPick = infos.begin();
        }

        publishIfChanged(toState(*displayPick, false));
    } catch (...) {
        publishIfChanged({});
    }
}

void MediaMonitor::publishIfChanged(MediaState state)
{
    Callback callback;
    {
        std::lock_guard lock(mutex_);
        if (current_.artist == state.artist && current_.title == state.title &&
            current_.isPlaying == state.isPlaying && current_.sourceApp == state.sourceApp)
            return;

        current_ = std::move(state);
        callback = callback_;
    }

    if (callback)
        callback(current());
}

}
